ptvsd/socket.py

```diff
@@ -45,6 +61,80 @@ def ignored_errno(*ignored):
             raise
 
 
+class KeepAlive(namedtuple('KeepAlive', 'interval idle, maxfails')):
+    """TCP keep-alive settings."""
+
+    INTERVAL = 3  # seconds
+    IDLE = 1  # seconds after idle
+    MAX_FAILS = 5
+
+    @classmethod
+    def from_raw(cls, raw):
+        """Return the corresponding KeepAlive."""
+        if raw is None:
+            return None
+        elif isinstance(raw, cls):
+            return raw
+        elif isinstance(raw, (str, int, float)):
+            return cls(raw)
+        else:
+            try:
+                raw = dict(raw)
+            except TypeError:
+                return cls(*raw)
+            else:
+                return cls(**raw)
+
+    def __new__(cls, interval=None, idle=None, maxfails=None):
+        self = super(KeepAlive, cls).__new__(
+            cls,
+            float(interval) if interval or interval == 0 else cls.INTERVAL,
+            float(idle) if idle or idle == 0 else cls.IDLE,
+            float(maxfails) if maxfails or maxfails == 0 else cls.MAX_FAILS,
+        )
+        return self
+
+    def apply(self, sock):
+        """Set the keepalive values on the socket."""
+        sock.setsockopt(socket.SOL_SOCKET,
+                        socket.SO_KEEPALIVE,
+                        1)
+        try:
```

Comment:
Might be worth moving `self.` attribute access out of the try-block here (i.e. assign to locals first), to avoid potential issues with the wrong `AttributeError` being caught and swallowed if e.g. someone renames an attribute.

Reply:
It shouldn't be an issue since the property would have to be actively removed.  However, I don't mind pulling those lookups out. :)  Regardless, I'm glad you made me look at this code as I found a typo. :)